src/frontend/src/scheduler/hummock_snapshot_manager.rs

```diff
@@ -58,27 +58,43 @@ impl HummockSnapshotManager {
     }
 
     pub async fn unpin_snapshot(&self, epoch: u64, query_id: &QueryId) -> Result<()> {
-        let local_count = async {
+        let min_epoch = async {
+            // Decrease the ref count of corresponding epoch. If all last pinned epoch is dropped,
+            // mark as outdated.
             let mut core_guard = self.core.lock().await;
             let query_ids = core_guard.epoch_to_query_ids.get_mut(&epoch);
             if let Some(query_ids) = query_ids {
                 query_ids.remove(query_id);
+                if query_ids.is_empty() && epoch == core_guard.last_pinned {
+                    core_guard.is_outdated = true;
+                }
+            }
+
+            // Iterate for all query and calculate the min epoch.
+            let mut min_epoch = None;
+            for (epoch, query_ids) in &core_guard.epoch_to_query_ids {
                 if query_ids.is_empty() {
-                    core_guard.epoch_to_query_ids.remove(&epoch);
-                    if epoch == core_guard.last_pinned {
-                        core_guard.is_outdated = true;
-                    }
-                    return true;
+                    min_epoch = Some(*epoch);
```

Comment:
The reason we can ignore those epochs/snapshots smaller than `min_epoch` but with queries running on it is because these queries must have acquired the HummockVersion they need to run on?

Reply:
Now we use a BTreeMap, so we are always pick the smallest epoch without epochs running on it?

Reply:
For example, epoch 1 still got queries running, epoch 2 has none in this call and is `min_epoch` in this case. 

By calling `unpin_snapshot_before` with epoch 2, we also unpin epoch 1. Is it possible that some queries with epoch 1 haven't been initialized and cannot acquire the SSTs needed as they just got GCed(due to epoch 1 unpinned)?

Maybe I miss some about the big picture of the new design.

Reply:
Oh indeed. I should only check the min epoch instead of doing a for loop. It's a logic error and I think you are right. 

Minds get messy..